codegen/src/main/java/org/web3j/codegen/unit/gen/CompilerClassLoader.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 Web3 Labs Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.web3j.codegen.unit.gen;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.tools.JavaCompiler;
+import javax.tools.JavaCompiler.CompilationTask;
+import javax.tools.JavaFileObject;
+import javax.tools.ToolProvider;
+
+/** Class loader with compilation capabilities. */
+class CompilerClassLoader extends ClassLoader {
+
+    private final File outputDir;
+    private final URL[] urls;
+
+    /**
+     * Creates a class loader from the given source URLs.
+     *
+     * @param outputDir Directory where classes will be compiled.
+     * @param urls Classpath URLs to compile the Java sources.
+     */
+    CompilerClassLoader(final File outputDir, final URL... urls) {
+        super(CompilerClassLoader.class.getClassLoader());
+        this.outputDir = outputDir;
+        this.urls = urls;
+    }
+
+    @Override
+    protected Class<?> findClass(final String name) throws ClassNotFoundException {
+        return compileClass(name)
+                .flatMap(this::readBytes)
+                .map(bytes -> defineClass(name, bytes, 0, bytes.length))
+                .orElseThrow(() -> new ClassNotFoundException(name));
+    }
+
+    private Optional<File> compileClass(final String name) {
```

Comment:
This seems similar to some of the logic we have in the codegen test classes for compiling generated code, please can we ensure we're not duplicating it here.